modules/index-rebuild/src/main/java/org/opencastproject/index/rebuild/IndexRebuildService.java

```diff
@@ -0,0 +1,228 @@
+  private final Map<IndexRebuildService.Service, IndexProducer> indexProducers = new HashMap<>();
+  private ServiceRegistration<?> serviceRegistration = null;
+
+  @Override
+  public void start(BundleContext bundleContext) throws Exception {
+
+    // check if there are already indexProducers available
+    ServiceReference<?>[] serviceReferences = bundleContext.getAllServiceReferences(IndexProducer.class.getName(),
+            null);
+    if (serviceReferences != null) {
+      for (ServiceReference<?> serviceReference : serviceReferences) {
+        addIndexProducer((IndexProducer) bundleContext.getService(serviceReference));
+      }
+    }
+
+    // all available?
+    if (indexProducers.size() == IndexRebuildService.Service.values().length) {
+      registerIndexRebuildService(bundleContext);
+    }
+
+    // listen to changes in availability
+    bundleContext.addServiceListener(new IndexProducerListener(bundleContext),
+            "(objectClass=" + IndexProducer.class.getName() + ")");
+  }
+
+  @Override
+  public void stop(BundleContext bundleContext) throws Exception {
+    unregisterIndexRebuildService();
+  }
+
+  /**
+   * Recreate the index from all of the services that provide data for it.
+   *
+   * @param index
+   *           The index to rebuild.
+   *
+   * @throws IOException
+   *           Thrown if the index cannot be cleared.
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  public synchronized void recreateIndex(AbstractSearchIndex index)
+          throws IOException, IndexRebuildException {
+    index.clear();
+    for (IndexRebuildService.Service service: IndexRebuildService.Service.values()) {
+      recreateService(index, service);
+    }
+  }
+
+  /**
+   * Recreate the index from a specific service that provides data for it.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param serviceName
+   *           The name of the {@link Service}.
+   *
+   * @throws IllegalArgumentException
+   *           Thrown if the service doesn't exist.
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  public synchronized void recreateIndex(AbstractSearchIndex index, String serviceName)
+          throws IllegalArgumentException, IndexRebuildException {
+    IndexRebuildService.Service service = IndexRebuildService.Service.valueOf(serviceName);
+    recreateService(index, service);
+  }
+
+  /**
+   * Trigger repopulation of the specified index with data from the specified service.
+   *
+   * @param index
+   *           The index to rebuild.
+   * @param service
+   *          The {@link IndexRebuildService.Service} to re-add data from.
+   *
+   * @throws IndexRebuildException
+   *           Thrown if the index rebuild failed.
+   */
+  private void recreateService(AbstractSearchIndex index, IndexRebuildService.Service service)
```

Comment:
This does not recreate a service, does it?
The names here are confusing.

Reply:
I know! In my defense, I didn't invent those, they were already like this (just in another place). I was too lazy to rename them, but maybe I should...